<html lang="en">
  <head>
    <meta charset="UTF-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<meta http-equiv="X-UA-Compatible" content="ie=edge" />
<title>A brief compasison of GSMMs reconstruction -
    阿宾的BLOG</title>
<link rel="shortcut icon" href="https://xiaobin-phd.github.io/favicon.ico" />
<link
  href="https://unpkg.com/remixicon@2.2.0/fonts/remixicon.css"
  rel="stylesheet" />
<link
  rel="stylesheet"
  href="https://unpkg.com/animate.css@3.7.2/animate.min.css" />
<link
  rel="stylesheet"
  href="https://xiaobin-phd.github.io/media/css/tailwind.css" />
<link rel="stylesheet" href="https://xiaobin-phd.github.io/styles/main.css" />
<link
  rel="alternate"
  type="application/atom+xml"
  title="A brief compasison of GSMMs reconstruction -
    阿宾的BLOG - Atom Feed"
  href="https://xiaobin-phd.github.io/atom.xml" />



    <meta name="description" content="1. AuReMe (Automatic Reconstruction of Metabolic Models，2018)
# Description
good traceability of the whole reconstructio..." />
    <meta
      property="og:title"
      content="A brief compasison of GSMMs reconstruction - 阿宾的BLOG" />
    <meta property="og:description" content="1. AuReMe (Automatic Reconstruction of Metabolic Models，2018)
# Description
good traceability of the whole reconstructio..." />
    <meta property="og:type" content="articles" />
    <meta property="og:url" content="https://xiaobin-phd.github.io/post/a-brief-compasison-of-gsmms-reconstruction/" />
    <meta
      property="og:image"
      content="https://xiaobin-phd.github.io/images/avatar.png" />
    <meta property="og:image:height" content="630" />
    <meta property="og:image:width" content="1200" />
    <meta
      name="twitter:title"
      content="A brief compasison of GSMMs reconstruction - 阿宾的BLOG" />
    <meta name="twitter:description" content="1. AuReMe (Automatic Reconstruction of Metabolic Models，2018)
# Description
good traceability of the whole reconstructio..." />
    <meta name="twitter:card" content="summary_large_image" />
    <link rel="canonical" href="https://xiaobin-phd.github.io/post/a-brief-compasison-of-gsmms-reconstruction/" />

    <link
      rel="stylesheet"
      href="https://unpkg.com/photoswipe@4.1.3/dist/photoswipe.css" />
    <link
      rel="stylesheet"
      href="https://unpkg.com/photoswipe@4.1.3/dist/default-skin/default-skin.css" />

     
  </head>

  <body>
    <div class="antialiased flex flex-col min-h-screen" id="app">
      <a
        href="https://xiaobin-phd.github.io"
        class="fixed top-0 left-0 mt-4 bg-black text-white dark:text-gray-700 dark:bg-yellow-50 dark:hover:bg-black dark:hover:text-white inline-flex p-2 pl-8 hover:text-gray-700 hover:bg-yellow-50 font-bold z-10 transition-fast animated fadeInLeft">
        阿宾的BLOG
      </a>
      <div class="max-w-4xl w-full mx-auto">
        <div
          class="shadow-box bg-white dark:bg-gray-600 rounded-lg pt-32 md:pt-64 px-4 md:px-8 pb-8 animated fadeIn mb-8">
          <h1
            class="text-5xl font-semibold leading-normal pb-8 mb-8 border-b-8 border-gray-700">
            A brief compasison of GSMMs reconstruction
          </h1>
          
          <div class="mb-8 flex flex-wrap">
            <div class="text-gray-400 text-sm mr-4">
              2024-06-18 · 4 min read
            </div>
            
            <a
              href="https://xiaobin-phd.github.io/tag/nphTmqWO-/"
              class="text-gray-700 text-sm border-b-2 border-dotted border-gray-200 hover:border-gray-600 transition-all duration-100 inline-flex mr-2">
              <i class="ri-hashtag"></i>
              BIOINFORMATICS
            </a>
            
          </div>
          <div class="markdown mb-8" v-pre><h3 id="1-aureme-automatic-reconstruction-of-metabolic-models2018">1. AuReMe (Automatic Reconstruction of Metabolic Models，2018)</h3>
<pre><code class="language-bash"># Description
good traceability of the whole reconstruction process
offer docker image
creates GSMMs with a template-based algorithm
incorporate information from MetaCyc and BIGG

# reference
DOI: https://doi.org/10.1093/nar/gkv1049
</code></pre>
<h3 id="2-carveme-2018">2. CarveMe (2018)</h3>
<pre><code class="language-bash"># Description
command-line python-based tool
use for Flux Balance Analysis (FBA) in a few minutes
create models from a BIGG-based manually curated universal template
own gap-filling algorithm
prioritize the incorporation into the network of reactions with higher genetic evidence

# reference
DOI: https://doi.org/10.1093/nar/gky537
</code></pre>
<h3 id="3-metadraft-2018">3. MetaDraft (2018)</h3>
<pre><code class="language-bash"># Description
Python-based user-friendly software
create GSMMs from manually curated models(BIGG or any other)
users can set priorities when multiple templates match
support the SBML Level 3

# reference
DOI: https://doi.org/10.1371/journal.pone.0173183
DOI: 10.5281/zenodo.2398336
</code></pre>
<h3 id="4-raven-version2-reconstruction-analysis-and-visualization-of-metabolic-networks-2018">4. RAVEN version2 (Reconstruction, Analysis and Visualization of Metabolic Networks, 2018)</h3>
<pre><code class="language-bash"># Description
runs in MATLAB is compatible with COBRA Toolbox v3
allow users reconstruct GSMMs from KEGG, MetaCys(transporters and spontaneous reactions) and template models
integrate an algorithm that merges network from two databases

# reference
DOI: https://doi.org/10.1371/journal.pcbi.1006541
</code></pre>
<h3 id="5-modelseed-version-22-2018">5. ModelSEED version 2.2 (2018)</h3>
<pre><code class="language-bash"># Description
web resource for reconstruction of GSMMs (microorganisms and plants)
use RAST for genome annotation (first step)
users can select or create a medium for gap-filling
allow model creation in under 10 minutes (including annotation) 
provide reaction and metabolite aliases in other databases

# reference
DOI: https://doi.org/10.1038/nbt.1672
</code></pre>
<h3 id="6-pathway-tools-version-220-2018">6. Pathway Tools version 22.0 (2018)</h3>
<pre><code class="language-bash"># Description
users can interactively explore, visualize and edit different components
network can be visualized using Cellular Overview diagrams
experimental data like gene expression can be color-mapped based on expression levels

# reference
DOI: https://doi.org/10.1093/bib/bbv079
</code></pre>
<h3 id="7-merlin-version-38-2018-flexible">7. Merlin version 3.8 (2018) flexible</h3>
<pre><code class="language-bash">java application based on the KEGG database
re-annotation of genomes through the online service of BLAST (EBI) or HMMER
flexible adjust parameters (expected value threshold and maximum number) of annotation algorithms
the interface enables comparing gene function agreement between annotation and UniProt

# reference
DOI: https://doi.org/10.1093/nar/gkv294
</code></pre>
<p><strong>Table 1 List of selected genome-scale metabolic reconstruction tools and their main features</strong></p>
<table>
<thead>
<tr>
<th style="text-align:left">Reconstruction tool</th>
<th style="text-align:left">Mapping method</th>
<th style="text-align:left">Reactions are inherited from</th>
<th style="text-align:left">Associated databases</th>
<th style="text-align:left">Version</th>
<th style="text-align:left">Type of software</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">AuReMe</td>
<td style="text-align:left">Pantograph (Inparanoid and OrthoMCL)</td>
<td style="text-align:left">Template model(s)</td>
<td style="text-align:left">BIGG-MetaCyc</td>
<td style="text-align:left">1.2.4</td>
<td style="text-align:left">Command line</td>
</tr>
<tr>
<td style="text-align:left">CarveMe</td>
<td style="text-align:left">Diamond, eggNOG-mappera</td>
<td style="text-align:left">Template model</td>
<td style="text-align:left">BIGG</td>
<td style="text-align:left">1.2.1</td>
<td style="text-align:left">Command line</td>
</tr>
<tr>
<td style="text-align:left">Merlin</td>
<td style="text-align:left">Mapping from annotation with BLAST or HMMER</td>
<td style="text-align:left">Database</td>
<td style="text-align:left">KEGG</td>
<td style="text-align:left">3.8</td>
<td style="text-align:left">Standalone interface</td>
</tr>
<tr>
<td style="text-align:left">MetaDraft</td>
<td style="text-align:left">Autograph (Inparanoid)</td>
<td style="text-align:left">Template model(s)</td>
<td style="text-align:left">BIGG</td>
<td style="text-align:left">0.9.2</td>
<td style="text-align:left">Standalone interface</td>
</tr>
<tr>
<td style="text-align:left">ModelSEED</td>
<td style="text-align:left">Annotation ontology map from RAST data</td>
<td style="text-align:left">Template model</td>
<td style="text-align:left">ModelSEED</td>
<td style="text-align:left">2.2–2.4</td>
<td style="text-align:left">Online service</td>
</tr>
<tr>
<td style="text-align:left">Pathway Tools</td>
<td style="text-align:left">Pathologic</td>
<td style="text-align:left">Database</td>
<td style="text-align:left">MetaCyc</td>
<td style="text-align:left">22.0</td>
<td style="text-align:left">Standalone interface</td>
</tr>
<tr>
<td style="text-align:left">RAVEN</td>
<td style="text-align:left">Autograph-type method from BLASTP and Bidirectional BLASTPb</td>
<td style="text-align:left">Database- Template model(s)</td>
<td style="text-align:left">KEGG-MetaCyc</td>
<td style="text-align:left">2.0.1</td>
<td style="text-align:left">Command line</td>
</tr>
</tbody>
</table>
<p>Reference: Mendoza SN, Olivier BG, Molenaar D, Teusink B. A systematic assessment of current genome-scale metabolic reconstruction tools. Genome Biol. 2019 Aug 7;20(1):158. doi: 10.1186/s13059-019-1769-1. PMID: 31391098; PMCID: PMC6685185.</p>
</div>
          <!-- Share to Twitter, Weibo, Telegram -->
          <div class="flex items-center">
            <div class="mr-4 flex items-center">
              <i class="ri-share-forward-line text-gray-500"></i>
            </div>
            <div
              class="px-4 cursor-pointer text-blue-500 hover:bg-blue-100 dark:hover:bg-gray-600 inline-flex"
              @click="shareToTwitter">
              <i class="ri-twitter-line"></i>
            </div>
            <div
              class="px-4 cursor-pointer text-red-500 hover:bg-red-100 dark:hover:bg-gray-600 inline-flex"
              @click="shareToWeibo">
              <i class="ri-weibo-line"></i>
            </div>
            <div
              class="px-4 cursor-pointer text-indigo-500 hover:bg-indigo-100 dark:hover:bg-gray-600 inline-flex"
              @click="shareToTelegram">
              <i class="ri-telegram-line"></i>
            </div>
          </div>
        </div>

          <footer class="py-12 text-center px-4 md:px-0" v-pre>
  The best way to predict the future is to create it！
</footer>

      </div>

      <!-- TOC Container -->
      <div
        class="fixed right-0 bottom-0 mb-16 mr-4 shadow w-8 h-8 rounded-full flex justify-center items-center z-10 cursor-pointer bg-white dark:bg-gray-500 dark:text-gray-200 hover:shadow-lg transition-all animated fadeInRight"
        @click="showToc = true">
        <i class="ri-file-list-line"></i>
      </div>

      <div
        class="fixed right-0 top-0 bottom-0 overflow-y-auto w-64 bg-white dark:bg-gray-800 p-4 border-l border-gray-100 dark:border-gray-600 z-10 transition-fast"
        :class="{ '-mr-64': !showToc }">
        <div class="flex mb-4 justify-end">
          <div
            class="w-8 h-8 inline-flex justify-center items-center rounded-full cursor-pointer hover:bg-gray-200 dark:hover:bg-gray-600 transition-fast"
            @click="showToc = false">
            <i class="ri-close-line text-lg"></i>
          </div>
        </div>
        <div class="post-toc-container"><ul class="markdownIt-TOC">
<li>
<ul>
<li>
<ul>
<li><a href="#1-aureme-automatic-reconstruction-of-metabolic-models2018">1. AuReMe (Automatic Reconstruction of Metabolic Models，2018)</a></li>
<li><a href="#2-carveme-2018">2. CarveMe (2018)</a></li>
<li><a href="#3-metadraft-2018">3. MetaDraft (2018)</a></li>
<li><a href="#4-raven-version2-reconstruction-analysis-and-visualization-of-metabolic-networks-2018">4. RAVEN version2 (Reconstruction, Analysis and Visualization of Metabolic Networks, 2018)</a></li>
<li><a href="#5-modelseed-version-22-2018">5. ModelSEED version 2.2 (2018)</a></li>
<li><a href="#6-pathway-tools-version-220-2018">6. Pathway Tools version 22.0 (2018)</a></li>
<li><a href="#7-merlin-version-38-2018-flexible">7. Merlin version 3.8 (2018) flexible</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
      </div>

      <!-- Back to top -->
      <div
        class="fixed right-0 bottom-0 mb-4 mr-4 shadow w-8 h-8 rounded-full flex justify-center items-center z-10 cursor-pointer bg-white hover:shadow-lg transition-all dark:bg-gray-500 dark:text-gray-200"
        @click="backToUp"
        v-show="scrolled">
        <i class="ri-arrow-up-line"></i>
      </div>
    </div>

    <!-- Root element of PhotoSwipe. Must have class pswp. -->
<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">
  <!-- Background of PhotoSwipe.
        It's a separate element as animating opacity is faster than rgba(). -->
  <div class="pswp__bg"></div>
  <!-- Slides wrapper with overflow:hidden. -->
  <div class="pswp__scroll-wrap">
    <!-- Container that holds slides.
            PhotoSwipe keeps only 3 of them in the DOM to save memory.
            Don't modify these 3 pswp__item elements, data is added later on. -->
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    <!-- Default (PhotoSwipeUI_Default) interface on top of sliding area. Can be changed. -->
    <div class="pswp__ui pswp__ui--hidden">
      <div class="pswp__top-bar">
        <!--  Controls are self-explanatory. Order can be changed. -->
        <div class="pswp__counter"></div>
        <button
          class="pswp__button pswp__button--close"
          title="Close (Esc)"></button>
        <button
          class="pswp__button pswp__button--fs"
          title="Toggle fullscreen"></button>
        <button
          class="pswp__button pswp__button--zoom"
          title="Zoom in/out"></button>
        <!-- Preloader demo http://codepen.io/dimsemenov/pen/yyBWoR -->
        <!-- element will get class pswp__preloader--active when preloader is running -->
        <div class="pswp__preloader">
          <div class="pswp__preloader__icn">
            <div class="pswp__preloader__cut">
              <div class="pswp__preloader__donut"></div>
            </div>
          </div>
        </div>
      </div>
      <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
        <div class="pswp__share-tooltip"></div>
      </div>
      <button
        class="pswp__button pswp__button--arrow--left"
        title="Previous (arrow left)"></button>
      <button
        class="pswp__button pswp__button--arrow--right"
        title="Next (arrow right)"></button>
      <div class="pswp__caption">
        <div class="pswp__caption__center"></div>
      </div>
    </div>
  </div>
</div>


    <script src="https://unpkg.com/vue@2.7.14/dist/vue.min.js"></script>
    <script src="https://xiaobin-phd.github.io/media/scripts/main.js"></script>

    <!-- Code Highlight -->
    

    <script src="https://unpkg.com/photoswipe@4.1.3/dist/photoswipe.min.js"></script>
    <script src="https://unpkg.com/photoswipe@4.1.3/dist/photoswipe-ui-default.min.js"></script>
    <script>
      //拿到预览框架，也就是上面的html代码
      var pswpElement = document.querySelectorAll('.pswp')[0]
      //定义图片数组变量
      var imgitems
      /**
       * 用于显示预览界面
       * @param index 图片数组下标
       */
      function viewImg(index) {
        //其它选项这里不做过多阐述，详情见官网
        var pswpoptions = {
          index: parseInt(index, 10), // 开始幻灯片索引。0是第一张幻灯片。必须是整数，而不是字符串。
          bgOpacity: 0.7, // 背景透明度，0-1
          maxSpreadZoom: 3, // 缩放级别，不要太大
        }
        //初始化并打开PhotoSwipe，pswpElement对应上面预览框架，PhotoSwipeUI_Default为皮肤，imgitems为图片数组，pswpoptions为选项
        var gallery = new PhotoSwipe(
          pswpElement,
          PhotoSwipeUI_Default,
          imgitems,
          pswpoptions
        )
        gallery.init()
      }
      /**
       * 用于添加图片点击事件
       * @param img 图片元素
       * @param index 所属下标（在imgitems中的位置）
       */
      function addImgClick(img, index) {
        img.onclick = function () {
          viewImg(index)
        }
      }
      /**
       * 轮询所有图片，获取src、width、height等数据，加入imgitems，并给图片元素添加事件
       * 最好在onload中执行该方法，本站因放在最底部，所以直接初始化
       * 异步加载图片可在图片元素创建完成后调用此方法
       */
      function initImg() {
        //重置图片数组
        imgitems = []
        //查找class:markdown 下的所有img元素并遍历
        var imgs = document.querySelectorAll('.markdown img')
        for (var i = 0; i < imgs.length; i++) {
          var img = imgs[i]
          //本站相册初始为loading图片，真实图片放在data-src
          var ds = img.getAttribute('data-src')
          //创建image对象，用于获取图片宽高
          var imgtemp = new Image()
          //判断是否存在data-src
          if (ds != null && ds.length > 0) {
            imgtemp.src = ds
          } else {
            imgtemp.src = img.src
          }
          //判断是否存在缓存
          if (imgtemp.complete) {
            var imgobj = {
              src: imgtemp.src,
              w: imgtemp.width,
              h: imgtemp.height,
            }
            imgitems[i] = imgobj
            addImgClick(img, i)
          } else {
            console.log('进来了2')
            imgtemp.index = i
            imgtemp.img = img
            imgtemp.onload = function () {
              var imgobj = {
                src: this.src,
                w: this.width,
                h: this.height,
              }
              //不要使用push，因为onload前后顺序会不同
              imgitems[this.index] = imgobj
              //添加点击事件
              addImgClick(this.img, this.index)
            }
          }
        }
      }
      //初始化
      initImg()
    </script>
     
  </body>
</html>
